_posts/2019-10-07-strimzi-apache-kafka-and-tracing.md

```diff
@@ -0,0 +1,500 @@
+The Jaeger client is responsible for propagating the information about the spans to the Jaeger agent.
+From Jaeger agent the tracing information will be further passed to the Jaeger collector, ingester and query components.
+For the full description of the Jaeger architecture please see [Jaeger documentation](https://www.jaegertracing.io/docs/1.14/architecture/).
+
+The client doesn't usually send all traces to the agent but often only small percentage.
+A _sampling configuration_ can define how many traces will be sent to the agent.
+It can be configured either in the client directly or you can configure the client to get this configuration from the Jaeger agent.
+That will let you manage the sampling configuration centrally for all your applications.
+
+The OpenTracing project also contains a [_contributions_ repository](https://github.com/opentracing-contrib) which contains many integrations between the OpenTracing APIs and other projects.
+One such integration is for the Java based Apache Kafka clients and we will use it as well.
+
+## Consumer and Producer APIs
+
+OpenTracing has a [contributions project](https://github.com/opentracing-contrib/java-kafka-client) which already adds tracing support to the Consumer and Producer APIs.
+It will take care of managing the tracing spans, adding the trace IDs to the records when they are sent or reading them from the records which are received.
+It has two way how to enable tracing:
+* Using wrappers around the `KafkaConsumer` and `KafkaProducer` classes
+* Using `Interceptor` classes
+
+To use the wrapper classes, all you need to do is to create an instance of the `TracingKafkaProducer` or `TracingKafkaConsumer` class.
+You have to pass the original `KafkaProducer` or `KafkaConsumer` when creating the wrapper class.
+Once created, the wrapper class provides the same interface as the original Kafka classes and can be used in the same way.
+
+```java
+// Producer
+KafkaProducer producer = new KafkaProducer(props);
+TracingKafkaProducer tracingProducer = new TracingKafkaProducer(producer);
+tracingProducer.send(new ProducerRecord("my-topic","Hello world"));
+
+// Consumer
+KafkaConsumer consumer = new KafkaConsumer(props);
+TracingKafkaConsumer tracingConsumer = new TracingKafkaConsumer(consumer);
+tracingConsumer.subscribe(Collections.singletonList("my-topic"));
+tracingConsumer.poll(Long.MAX_VALUE);
+```
+
+To use the interceptors, all you need to do is to set the interceptor class in the producer or consumer configuration:
+
+```java
+// Producer
+props.put(ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingProducerInterceptor.class.getName());
+
+// Consumer
+props.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingConsumerInterceptor.class.getName());
+```
+
+You don't have to do anything else.
+You use the `KafkaProducer` and `KafkaConsumer` classes as always and the interceptors will take care of the rest.
```

Comment:
what are the main reasons to choose one way or the other? If any should we explain that to the reader?

Reply:
For me it is more or less question of preference. So i probably wouldn't go into details and let people choose.

Reply:
So it's worth mentioning that it's just a matter of preferences otherwise people could think that there are differences that you are not explaining.